# 1.0 Set up

## 1.1 Import neccessary libraries

In [1]:
# Data manipulation and analysis
import pandas as pd
import numpy as np

# Directories
from pathlib import Path

## 1.2 Paths

In [2]:
# Current directory
CURRENT_DIR = Path.cwd()

# Path to main folder
PROJECT_ROOT = CURRENT_DIR.parent

# Path to folders and files
RAW_EXCEL = PROJECT_ROOT/"data"/"raw"/"Online_Retail.xlsx"
RAW_CSV = PROJECT_ROOT/"data"/"processed"/"Online_Retail.csv"
PROCESSED = PROJECT_ROOT/"data"/"processed"

## 1.3 Convert Excel file to CSV

In [3]:
# Convert file from Excel to CSV only if CSV doesn't exist
if not RAW_CSV.is_file():
    df_excel = pd.read_excel(RAW_EXCEL, sheet_name="Online Retail", engine="openpyxl")
    
    # Save as CSV
    df_excel.to_csv(RAW_CSV, index=None, header=True)
    print(f"CSV file saved at {RAW_CSV}")
else:
    print(f"CSV file already exists at {RAW_CSV}")


CSV file already exists at c:\my_projects\GIT\retail_analysis\data\processed\Online_Retail.csv


## 1.4 Load dataset

In [4]:
# Load the dataset
original_df = pd.read_csv(RAW_CSV)

# First 5 rows of the dataset
original_df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


# 2.0 Initial analysis

In [5]:
# Shape of the dataset
print(f'Number of rows:{original_df.shape[0]}')
print(f'Number of columns:{original_df.shape[1]}')

Number of rows:541909
Number of columns:8


In [6]:
# General information
original_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [7]:
# Statistical summary
original_df.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


### Output summary
- The dataset has 541,909 rows and 8 columns.

- Description and CustomerID have null values.

- The dataset has 2 floats, 1 integer and 5 objects data types. Invoice date has object data type.

- There is a negative minimum value for Quantity and UnitPrice.

# 3.0 Data Preprocessing

## 3.1 Convert datatypes

In [8]:
# Convert Invoice date from object to datetime
cleaned_df = original_df.copy()
cleaned_df['InvoiceDate'] = pd.to_datetime(cleaned_df['InvoiceDate'])

# Convert CustomerId datatype from float to integer
cleaned_df['CustomerID'] = cleaned_df['CustomerID'].astype('Int64')

cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  Int64         
 7   Country      541909 non-null  object        
dtypes: Int64(1), datetime64[ns](1), float64(1), int64(1), object(4)
memory usage: 33.6+ MB


## 3.2 Create 'Total price' column

In [9]:
# Create Total price column
cleaned_df['TotalPrice'] = cleaned_df['UnitPrice'] * cleaned_df['Quantity']
cleaned_df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34


In [10]:
# Dataset general informtion
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  Int64         
 7   Country      541909 non-null  object        
 8   TotalPrice   541909 non-null  float64       
dtypes: Int64(1), datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 37.7+ MB


### Output summary
- Converted datatypes: Invoice date to Datetime from object and Customer ID to int64 from float64.
- Created column: TotalPrice. Datatype is float64

## 3.3 Duplicates

In [11]:
# Check for duplicated rows
duplicates = cleaned_df.duplicated().sum()
print(duplicates)

5268


In [12]:
# Duplicated entries
duplicated_records = cleaned_df[cleaned_df.duplicated(keep=False)] # keeps all duplicates including the first occurrence

# Group identical rows together
duplicated_records = duplicated_records.sort_values(by=duplicated_records.columns.tolist()) 
duplicated_records.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom,1.25
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom,1.25
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom,4.95
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom,4.95
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom,2.10
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom,2.10
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom,2.95
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom,2.95
578,536412,21448,12 DAISY PEGS IN WOOD BOX,1,2010-12-01 11:49:00,1.65,17920,United Kingdom,1.65
598,536412,21448,12 DAISY PEGS IN WOOD BOX,1,2010-12-01 11:49:00,1.65,17920,United Kingdom,1.65


In [13]:
# Remove duplicates
cleaned_df = cleaned_df.drop_duplicates()
cleaned_df.shape

(536641, 9)

### Output summary
- 5,268 duplicate rows were dropped.

## 3.4 Missing values

In [14]:
# Check for missing values
missing = cleaned_df.isnull().sum()

# Percentage of total rows missing for specific column.
missing_percentage = (missing /len(cleaned_df) )*100
missing_df = pd.DataFrame({'Missing Values': missing, '% of Total': missing_percentage})
pd.set_option('display.precision', 4)
print(missing_df)

             Missing Values  % of Total
InvoiceNo                 0      0.0000
StockCode                 0      0.0000
Description            1454      0.2709
Quantity                  0      0.0000
InvoiceDate               0      0.0000
UnitPrice                 0      0.0000
CustomerID           135037     25.1634
Country                   0      0.0000
TotalPrice                0      0.0000


### 3.4.1 Missing Customer Id

In [15]:
# Confirm CustomerId is 5-digits
# Create dataframe copy 
cleaned_df2 = cleaned_df.copy()

# Customer ids that are not 5 digits
cleaned_df2['CustomerID_str'] = cleaned_df2['CustomerID'].astype(str)
odd_customerid = cleaned_df2[~cleaned_df2['CustomerID_str'].astype(str).str.fullmatch(r'\d{5}')]
odd_customerid

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,CustomerID_str
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.00,<NA>,United Kingdom,0.00,<NA>
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom,2.51,<NA>
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom,5.02,<NA>
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,<NA>,United Kingdom,3.40,<NA>
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,<NA>,United Kingdom,3.32,<NA>
...,...,...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,2011-12-09 10:26:00,4.13,<NA>,United Kingdom,20.65,<NA>
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,2011-12-09 10:26:00,4.13,<NA>,United Kingdom,16.52,<NA>
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,2011-12-09 10:26:00,4.96,<NA>,United Kingdom,4.96,<NA>
541539,581498,85174,S/4 CACTI CANDLES,1,2011-12-09 10:26:00,10.79,<NA>,United Kingdom,10.79,<NA>


In [16]:
# Confirm the odd customer ids are the same as null customer ids
# Missing customer id
null_customerid = cleaned_df[cleaned_df['CustomerID'].isnull()]

# Confirm
odd_customerid.drop(columns='CustomerID_str').equals(null_customerid)

True

In [17]:
# Fill null customer id if invoice number has at least 1 customer id
# Groups rows by invoice No and fill missing customerIds forwads and backwards
cleaned_df = cleaned_df.copy()
cleaned_df["CustomerID"] = cleaned_df.groupby("InvoiceNo")["CustomerID"].transform(lambda x: x.ffill().bfill())

# Check for missing values
cleaned_df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135037
Country             0
TotalPrice          0
dtype: int64

In [18]:
# Invoices with all customerId missing, give new customer ID (InvoiceNo + I)
missing_id = cleaned_df['CustomerID'].isnull()
cleaned_df = cleaned_df.copy()
cleaned_df['CustomerID'] = cleaned_df['CustomerID'].astype('object')
cleaned_df.loc[missing_id, 'CustomerID'] = cleaned_df.loc[missing_id, 'InvoiceNo'].astype(str) + 'I'

# Check for missing values
cleaned_df.isnull().sum()

InvoiceNo         0
StockCode         0
Description    1454
Quantity          0
InvoiceDate       0
UnitPrice         0
CustomerID        0
Country           0
TotalPrice        0
dtype: int64

In [19]:
# Confirm filling of null customer ID
cleaned_df[cleaned_df['InvoiceNo'] == '536414']

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,536414I,United Kingdom,0.0


### Output summary
- There are 135,037 rows which do not have 5 digit Ids. Those are the ones with null IDs.
- The invoices were not asscociated to any customer Id so no customer Id was filled
- Missing customer Ids filled with InvoiceNo + I.

### 3.4.2 Missing Description

In [20]:
# Descriptions that are null
cleaned_df[cleaned_df['Description'].isnull()].head(15)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,536414I,United Kingdom,0.0
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,536545I,United Kingdom,0.0
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,536546I,United Kingdom,0.0
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,536547I,United Kingdom,0.0
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,536549I,United Kingdom,0.0
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,536550I,United Kingdom,0.0
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,536552I,United Kingdom,0.0
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,536553I,United Kingdom,0.0
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,536554I,United Kingdom,0.0
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,536589I,United Kingdom,-0.0


In [21]:
# Orders with negative quantity
neg_quantity = cleaned_df[cleaned_df['Quantity'] < 0]
neg_quantity.tail(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
540141,C581468,21314,SMALL GLASS HEART TRINKET POT,-10,2011-12-08 19:26:00,2.10,13599,United Kingdom,-21.00
540142,C581468,22098,BOUDOIR SQUARE TISSUE BOX,-12,2011-12-08 19:26:00,0.39,13599,United Kingdom,-4.68
540176,C581470,23084,RABBIT NIGHT LIGHT,-4,2011-12-08 19:28:00,2.08,17924,United Kingdom,-8.32
540422,C581484,23843,"PAPER CRAFT , LITTLE BIRDIE",-80995,2011-12-09 09:27:00,2.08,16446,United Kingdom,-168469.60
540448,C581490,22178,VICTORIAN GLASS HANGING T-LIGHT,-12,2011-12-09 09:57:00,1.95,14397,United Kingdom,-23.40
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,2011-12-09 09:57:00,0.83,14397,United Kingdom,-9.13
541541,C581499,M,Manual,-1,2011-12-09 10:28:00,224.69,15498,United Kingdom,-224.69
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,2011-12-09 11:57:00,10.95,15311,United Kingdom,-54.75
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,2011-12-09 11:58:00,1.25,17315,United Kingdom,-1.25
541717,C581569,20979,36 PENCILS TUBE RED RETROSPOT,-5,2011-12-09 11:58:00,1.25,17315,United Kingdom,-6.25


In [22]:
# Shape of dataset with negative quantities
neg_quantity.shape

(10587, 9)

In [23]:
# Products with negative quantity and have 'C' in invoice number (cancelled order)
neg_cancelled = neg_quantity[(neg_quantity['InvoiceNo'].astype(str).str.startswith('C'))]
neg_cancelled.shape

(9251, 9)

In [24]:
# If Invoice No starts with letter 'c', it indicates a cancellation
# Products from the dataset that are cancelled
cancelled_orders = cleaned_df[cleaned_df['InvoiceNo'].astype(str).str.startswith('C')]
cancelled_orders.shape

(9251, 9)

In [25]:
# Confirm neg_cancelled is similar to cancelled orders
cancelled_orders.equals(neg_cancelled)

True

In [26]:
# Products with negative quantity but don't have 'C' in invoice number
neg_notcancelled = neg_quantity[~(neg_quantity['InvoiceNo'].astype(str).str.startswith('C'))]
neg_notcancelled

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,536589I,United Kingdom,-0.0
4347,536764,84952C,NaN,-38,2010-12-02 14:42:00,0.0,536764I,United Kingdom,-0.0
7188,536996,22712,NaN,-20,2010-12-03 15:30:00,0.0,536996I,United Kingdom,-0.0
7189,536997,22028,NaN,-20,2010-12-03 15:30:00,0.0,536997I,United Kingdom,-0.0
7190,536998,85067,NaN,-6,2010-12-03 15:30:00,0.0,536998I,United Kingdom,-0.0
...,...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,2011-12-07 18:36:00,0.0,581210I,United Kingdom,-0.0
535335,581212,22578,lost,-1050,2011-12-07 18:38:00,0.0,581212I,United Kingdom,-0.0
535336,581213,22576,check,-30,2011-12-07 18:38:00,0.0,581213I,United Kingdom,-0.0
536908,581226,23090,missing,-338,2011-12-08 09:56:00,0.0,581226I,United Kingdom,-0.0


In [27]:
# Descriptions of products with zero unit price and negtive quantity 
zero_negative = cleaned_df[(cleaned_df['UnitPrice'] == 0) & (cleaned_df['Quantity'] < 0)]
zero_negative['Description'].unique()

array([nan, '?', 'check', 'damages', 'faulty', 'Dotcom sales',
       'reverse 21/5/10 adjustment', 'mouldy, thrown away.', 'counted',
       'Given away', 'Dotcom', 'label mix up', 'samples/damages',
       'thrown away', 'incorrectly made-thrown away.', 'showroom', 'MIA',
       'Dotcom set', 'wrongly sold as sets', 'Amazon sold sets',
       'dotcom sold sets', 'wrongly sold sets', '? sold as sets?',
       '?sold as sets?', 'Thrown away.', 'damages/display',
       'damaged stock', 'broken', 'throw away', 'wrong barcode (22467)',
       'wrong barcode', 'barcode problem', '?lost',
       "thrown away-can't sell.", "thrown away-can't sell", 'damages?',
       're dotcom quick fix.', "Dotcom sold in 6's", 'sold in set?',
       'cracked', 'sold as 22467', 'Damaged',
       'mystery! Only ever imported 1800',
       'MERCHANT CHANDLER CREDIT ERROR, STO', 'POSSIBLE DAMAGES OR LOST?',
       'damaged', 'DAMAGED', 'Display', 'Missing', 'wrong code?',
       'wrong code', 'adjust', 'crush

In [28]:
# Products that are missing, damaged, had errors etc
product_error = zero_negative.copy()
product_error

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,536589I,United Kingdom,-0.0
4347,536764,84952C,NaN,-38,2010-12-02 14:42:00,0.0,536764I,United Kingdom,-0.0
7188,536996,22712,NaN,-20,2010-12-03 15:30:00,0.0,536996I,United Kingdom,-0.0
7189,536997,22028,NaN,-20,2010-12-03 15:30:00,0.0,536997I,United Kingdom,-0.0
7190,536998,85067,NaN,-6,2010-12-03 15:30:00,0.0,536998I,United Kingdom,-0.0
...,...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,2011-12-07 18:36:00,0.0,581210I,United Kingdom,-0.0
535335,581212,22578,lost,-1050,2011-12-07 18:38:00,0.0,581212I,United Kingdom,-0.0
535336,581213,22576,check,-30,2011-12-07 18:38:00,0.0,581213I,United Kingdom,-0.0
536908,581226,23090,missing,-338,2011-12-08 09:56:00,0.0,581226I,United Kingdom,-0.0


In [29]:
# Confirm neg_notcancelled is similar to product error
product_error.equals(neg_notcancelled)

True

In [30]:
# Drop the products with errors from dataset
cleaned_df = cleaned_df.drop(product_error.index)
cleaned_df.shape

(535305, 9)

In [31]:
# Drop cancelled orders
cleaned_df = cleaned_df.drop(cancelled_orders.index)
cleaned_df.shape

(526054, 9)

### Output summary
- The missing descriptions had products with negative quantities.

- The rows removed from full dataset are those with:
 - negative quantities and invoices that did not start with 'C' and 0 unit price. These are products with errors eg damaged, wet, lost etc
    - Products with errors do not have a standard description eg damaged stock, damaged, damages
 - negative quantities and invoices that start with 'C'. These are cancelled orders

In [32]:
# Map stockcode to description
mapping = original_df.dropna(subset=['Description']).drop_duplicates('StockCode').set_index('StockCode')['Description']

mapping

StockCode
85123A     WHITE HANGING HEART T-LIGHT HOLDER
71053                     WHITE METAL LANTERN
84406B         CREAM CUPID HEARTS COAT HANGER
84029G    KNITTED UNION FLAG HOT WATER BOTTLE
84029E         RED WOOLLY HOTTIE WHITE HEART.
                         ...                 
85179a                GREEN BITTY LIGHT CHAIN
23617     SET 10 CARDS SWIRLY XMAS TREE 17104
90214U              LETTER "U" BLING KEY RING
47591b               SCOTTIES CHILDRENS APRON
23843             PAPER CRAFT , LITTLE BIRDIE
Name: Description, Length: 3958, dtype: object

In [33]:
# Fill missing descriptions using mapping
cleaned_df.loc[cleaned_df['Description'].isnull(), 'Description'] = (
    cleaned_df.loc[cleaned_df['Description'].isnull(), 'StockCode'].map(mapping)
)

cleaned_df.isnull().sum()

InvoiceNo       0
StockCode       0
Description    15
Quantity        0
InvoiceDate     0
UnitPrice       0
CustomerID      0
Country         0
TotalPrice      0
dtype: int64

In [34]:
# Stockcodes with no descriptions
no_description = cleaned_df.loc[cleaned_df['Description'].isnull(), 'StockCode'].unique()
no_description

array(['21134', '85226A', '85044', '20950', '84670', '35951', '20849',
       '72803B', '85018C', '62095B', '72814', '84247C', '90042B',
       '35592T', '84971L'], dtype=object)

In [35]:
# Check entries left with null descriptions
cleaned_df[cleaned_df['Description'].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,536545I,United Kingdom,0.0
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,536549I,United Kingdom,0.0
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,536550I,United Kingdom,0.0
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,536552I,United Kingdom,0.0
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,536554I,United Kingdom,0.0
7187,536995,35951,NaN,57,2010-12-03 15:29:00,0.0,536995I,United Kingdom,0.0
19628,537875,20849,NaN,1,2010-12-08 18:08:00,0.0,537875I,United Kingdom,0.0
19631,537878,72803B,NaN,1,2010-12-08 18:09:00,0.0,537878I,United Kingdom,0.0
21782,538133,85018C,NaN,3,2010-12-09 15:56:00,0.0,538133I,United Kingdom,0.0
21786,538137,62095B,NaN,2,2010-12-09 15:57:00,0.0,538137I,United Kingdom,0.0


In [36]:
# Stockcodes without descriptions that appear more than once
#cleaned_df.loc[cleaned_df['Description'].isnull(), 'StockCode'].value_counts()
# Rows where description is null
null_desc = cleaned_df[cleaned_df['Description'].isnull()]

# Stockcode counts of of null desc
code_counts = null_desc['StockCode'].value_counts()

# Stockcode counts appearing more than once 
code_counts = code_counts[code_counts>1]

code_counts


Series([], Name: count, dtype: int64)

### Output summary
- 15 null Descriptions are left after mapping stock code to existing descriptions.
- The codes remaining without desriptions appear once each. So there are 15 unique stockcodes with no descriptions

In [37]:
# Turn  Stock codes with no description to dataframe
nodescription_df = pd.DataFrame({'StockCode': no_description})
nodescription_df

,StockCode
0,21134
1,85226A
2,85044
3,20950
4,84670
5,35951
6,20849
7,72803B
8,85018C
9,62095B


In [38]:
# Function to create pattern for stock code. D for numbers and L for letters
def get_pattern(code_numbers):
    code_numbers = str(code_numbers).strip().upper()
    pattern = ''
    for char in code_numbers:
        if char.isdigit():
            pattern += 'D'
        elif char.isalpha():
            pattern += 'L'
        else:
            pattern += '?'
    return pattern

nodescription_df['Pattern'] = nodescription_df['StockCode'].apply(get_pattern)


In [39]:
# categorize the patterns
import re
def group_pattern(p):
    if set(p) == {'D'}:
        return 'Numeric'
    elif re.fullmatch(r'D+L+', p):
        return 'Letter at end'
    elif re.fullmatch(r'L+D+', p):
        return 'Letter at start'
    elif re.fullmatch(r'[DL]+', p):
        return 'Mixed'
    else:
        return 'Other'
        
    
    
nodescription_df['Category'] = nodescription_df['Pattern'].apply(group_pattern)

nodescription_df[['StockCode', 'Pattern', 'Category']].head(20)


,StockCode,Pattern,Category
0,21134,DDDDD,Numeric
1,85226A,DDDDDL,Letter at end
2,85044,DDDDD,Numeric
3,20950,DDDDD,Numeric
4,84670,DDDDD,Numeric
5,35951,DDDDD,Numeric
6,20849,DDDDD,Numeric
7,72803B,DDDDDL,Letter at end
8,85018C,DDDDDL,Letter at end
9,62095B,DDDDDL,Letter at end


In [40]:
# Number of stock codes per pattern group
nodescription_df['Pattern'].value_counts()

Pattern
DDDDDL    8
DDDDD     7
Name: count, dtype: int64

In [41]:
# exclude digit only stockcode from nodescription_df
nodescription_df2 = nodescription_df[nodescription_df['Pattern'] != 'DDDDD'].copy()

# remove letters to remain with numbers
nodescription_df2['Numeric'] = nodescription_df2['StockCode'].astype(str).str.replace(r'[A-za-z]', '', regex=True)

# set of all stockcodes in cleaned df
allcodes = set(original_df['StockCode'].astype(str))

# filter odd stock codes that have a numeric counterpart
nodescription_df2['NumericMatches'] = nodescription_df2['Numeric'].isin(allcodes)

# codes that have a match
matched_df = nodescription_df2[nodescription_df2['NumericMatches']][['StockCode', 'Numeric']].astype(str).drop_duplicates()

matched_df.head()


,StockCode,Numeric


In [42]:
# Drop the remaining null values
cleaned_df = cleaned_df.dropna()
cleaned_df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
TotalPrice     0
dtype: int64

In [43]:
# Products with zero unit price
zero_price = cleaned_df[(cleaned_df['UnitPrice'] == 0)]
zero_price.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
622,536414,22139,RETROSPOT TEA SET CERAMIC 11 PC,56,2010-12-01 11:52:00,0.0,536414I,United Kingdom,0.0
1971,536546,22145,CHRISTMAS CRAFT HEART STOCKING,1,2010-12-01 14:33:00,0.0,536546I,United Kingdom,0.0
1972,536547,37509,NEW ENGLAND MUG W GIFT BOX,1,2010-12-01 14:33:00,0.0,536547I,United Kingdom,0.0
2025,536553,37461,FUNKY MONKEY MUG,3,2010-12-01 14:35:00,0.0,536553I,United Kingdom,0.0
4348,536765,84952C,MIRROR LOVE BIRD T-LIGHT HOLDER,19,2010-12-02 14:43:00,0.0,536765I,United Kingdom,0.0
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,536941I,United Kingdom,0.0
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,536942I,United Kingdom,0.0
7191,536999,21421,PORCELAIN ROSE LARGE,110,2010-12-03 15:32:00,0.0,536999I,United Kingdom,0.0
7194,537002,21651,HANGING GLASS ETCHED TEALIGHT,70,2010-12-03 15:33:00,0.0,537002I,United Kingdom,0.0
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 14:02:00,0.0,12647,Germany,0.0


In [44]:
# Descriptions of orders with zero unit price
zero_price['Description'].unique()

array(['RETROSPOT TEA SET CERAMIC 11 PC ',
       'CHRISTMAS CRAFT HEART STOCKING ', 'NEW ENGLAND MUG W GIFT BOX',
       'FUNKY MONKEY MUG', 'MIRROR LOVE BIRD T-LIGHT HOLDER', 'amazon',
       'PORCELAIN ROSE LARGE ', 'HANGING GLASS ETCHED TEALIGHT',
       'ROUND CAKE TIN VINTAGE GREEN', 'RECIPE BOX WITH METAL HEART',
       'ASSORTED CHEESE FRIDGE MAGNETS', 'ASSTD DESIGN BUBBLE GUM RING',
       'BEADED LOVE HEART JEWELLERY SET', 'WRAP CHRISTMAS VILLAGE',
       'SET/4 2 TONE EGG SHAPE MIXING BOWLS',
       'CREAM SWEETHEART LETTER RACK', 'ZINC WILLIE WINKIE  CANDLE STICK',
       'BOX OF 24 COCKTAIL PARASOLS', 'DOORMAT ENGLISH ROSE ',
       'DOORMAT 3 SMILEY CATS', 'GREEN REGENCY TEACUP AND SAUCER',
       'FRENCH BLUE METAL DOOR SIGN 7', 'FRENCH BLUE METAL DOOR SIGN 5',
       'FRENCH BLUE METAL DOOR SIGN 6', 'FRENCH BLUE METAL DOOR SIGN 4',
       'FRENCH BLUE METAL DOOR SIGN No', 'FRENCH BLUE METAL DOOR SIGN 8',
       'FRENCH BLUE METAL DOOR SIGN 1', 'RED KITCHEN SCALES',
    

In [45]:
# Shape of zero price
zero_price.shape

(1159, 9)

In [46]:
# Drop orders with zero unit price
cleaned_df = cleaned_df.drop(zero_price.index)
cleaned_df.shape

(524880, 9)

### Output summary
- For the remaining null descriptions, functions to create a pattern for the stock codes(D for numbers and L for letters) and to categorize those patterns were used.
- Stock codes with:
  5 numbers: 7 stock codes
  5 numbers with letter at the end : 8 stock codes
- The letters from the 8 stock codes with letters were removed and tried to match them with stock codes that had descriptions. However, there were no matches.
- All the rows remaining without descriptions also did not have Customer Ids originally and had unit price of 0. They were dropped.
- There were 1159 orders from the dataset that had zero unit price and quantity greater than zero might be promotional items or exchanges. They were dropped.

## 3.5 Invoice Number

In [47]:
# Confirm all Invoice no are 6-digit
odd_invoices = cleaned_df[~cleaned_df['InvoiceNo'].astype(str).str.fullmatch(r'\d{6}')]
odd_invoices

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
299982,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,A563185I,United Kingdom,11062.06
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,A563186I,United Kingdom,-11062.06
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,A563187I,United Kingdom,-11062.06


In [48]:
# Drop odd invoices(adjusted bad debt)
cleaned_df = cleaned_df.drop(odd_invoices.index)
cleaned_df.shape

(524877, 9)

### Output summary
- All invoice numbers that are not 6 digits are an adjustment for bad debt.
- They were dropped.

## 3.6 Unit price

In [49]:
# Rows that have negative unit price 
neg = cleaned_df[cleaned_df['UnitPrice'] < 0]
neg

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice


In [50]:
# Rows that have 0.0 unit price
zero = cleaned_df[cleaned_df['UnitPrice'] == 0]
zero

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice


### Output summary
- There are no rows with unit price zero or less than 0. They were dropped earlier

## 3.7 Stock code

In [51]:
# StockCode that is not is 5-digit
odd_stockcode = cleaned_df[~cleaned_df['StockCode'].astype(str).str.fullmatch(r'\d{5}')]
odd_stockcode

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
45,536370,POST,POSTAGE,3,2010-12-01 08:45:00,18.00,12583,France,54.00
...,...,...,...,...,...,...,...,...,...
541778,581578,84997C,CHILDRENS CUTLERY POLKADOT BLUE,8,2011-12-09 12:16:00,4.15,12713,Germany,33.20
541809,581579,85099C,JUMBO BAG BAROQUE BLACK WHITE,10,2011-12-09 12:19:00,1.79,17581,United Kingdom,17.90
541838,581580,84993A,75 GREEN PETIT FOUR CASES,2,2011-12-09 12:20:00,0.42,12748,United Kingdom,0.84
541844,581580,85049A,TRADITIONAL CHRISTMAS RIBBONS,1,2011-12-09 12:20:00,1.25,12748,United Kingdom,1.25


In [52]:
# Stockcodes that have letters only
stockcode_letters = cleaned_df[~cleaned_df['StockCode'].astype(str).str.contains(r'\d')]
stockcode_letters.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
45,536370,POST,POSTAGE,3,2010-12-01 08:45:00,18.00,12583,France,54.00
386,536403,POST,POSTAGE,1,2010-12-01 11:27:00,15.00,12791,Netherlands,15.00
1123,536527,POST,POSTAGE,1,2010-12-01 13:04:00,18.00,12662,Germany,18.00
1814,536544,DOT,DOTCOM POSTAGE,1,2010-12-01 14:32:00,569.77,536544I,United Kingdom,569.77
2239,536569,M,Manual,1,2010-12-01 15:35:00,1.25,16274,United Kingdom,1.25
2250,536569,M,Manual,1,2010-12-01 15:35:00,18.95,16274,United Kingdom,18.95
3041,536592,DOT,DOTCOM POSTAGE,1,2010-12-01 17:06:00,607.49,536592I,United Kingdom,607.49
4406,536779,BANK CHARGES,Bank Charges,1,2010-12-02 15:08:00,15.00,15823,United Kingdom,15.00
5073,536840,POST,POSTAGE,1,2010-12-02 18:27:00,18.00,12738,Germany,18.00
5258,536852,POST,POSTAGE,1,2010-12-03 09:51:00,18.00,12686,France,18.00


In [53]:
# Distinct StockCodes with letters
distinct_stockcode_letters = stockcode_letters['StockCode'].unique()
distinct_stockcode_letters

array(['POST', 'DOT', 'M', 'BANK CHARGES', 'AMAZONFEE', 'm', 'DCGSSBOY',
       'DCGSSGIRL', 'S', 'PADS'], dtype=object)

### Output summary
- There are stock codes with more and less than 5 digits.
- There are stock codes that have letters only.

## 3.8 Countries

In [54]:
# Countries where customers reside
countries = cleaned_df['Country'].unique().tolist()
countries.sort()
print(countries)

['Australia', 'Austria', 'Bahrain', 'Belgium', 'Brazil', 'Canada', 'Channel Islands', 'Cyprus', 'Czech Republic', 'Denmark', 'EIRE', 'European Community', 'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Iceland', 'Israel', 'Italy', 'Japan', 'Lebanon', 'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'RSA', 'Saudi Arabia', 'Singapore', 'Spain', 'Sweden', 'Switzerland', 'USA', 'United Arab Emirates', 'United Kingdom', 'Unspecified']


# 4.0 Final cleaned df

In [55]:
# Cleaned dataset shape
cleaned_df.shape

(524877, 9)

In [56]:
# Top 30 rows
cleaned_df.head(30)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,2010-12-01 08:26:00,7.65,17850,United Kingdom,15.30
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 08:26:00,4.25,17850,United Kingdom,25.50
7,536366,22633,HAND WARMER UNION JACK,6,2010-12-01 08:28:00,1.85,17850,United Kingdom,11.10
8,536366,22632,HAND WARMER RED POLKA DOT,6,2010-12-01 08:28:00,1.85,17850,United Kingdom,11.10
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,2010-12-01 08:34:00,1.69,13047,United Kingdom,54.08


In [57]:
# Save datasets
# Cleaned dataset 
cleaned_df.to_csv(PROCESSED/'cleaned_data.csv', index=False, float_format='%.2f', encoding='utf-8-sig', quoting=2)

# Missing/damaged products
product_error.to_csv(PROCESSED/'product_error.csv', index=False)

# Cancelled orders
cancelled_orders.to_csv(PROCESSED/'cancelled_orders.csv', index=False)

# Orders with zero unit price
zero_price.to_csv(PROCESSED/'zero_price.csv', index=False)
